### DISCUSSION TITLE
Proposal: Nullable var

### DISCUSSION THREAD

**Discussion by wojciechsura** [dotnet/csharplang#6985] [Language Ideas]
One of the most useful additions to C# language is `var` keyword, which simplifies declaration of variables, as long as compiler is able to deduce the type automatically.

I propose adding an option to quickly and automatically declare nullable type based on non-nullable one. For instance:

```csharp
Stack<(IReadOnlyList<BasePlanItem> folder, int index)> stack = new();
stack.Push(plan, -1);

var? current = stack.Peek();
// Instead of:
// (IReadOnlyList<BasePlanItem> folder, int index)? current = stack.Peek();
```

This is a small quality-of-life improvement, which may simplify lifes of developers, especially after introducing nullable reference types.

**Comment by theunrepentantgeek** [comment]
Why would you need `current` to be a wider type than the return type of `stack.Peek()`?

Either `stack.Peek()` already returns `(IReadOnlyList<BasePlanItem> folder, int index)` (in which case your variable can never be null) or it already returns `((IReadOnlyList<BasePlanItem> folder, int index)?` and you don't need `var?`.

(And, yes, I know that nullable reference types don't exist at runtime - but attributes on the implementation of `Peek()` have the same effect of declaring whether the return type is nullable or not.)

**Comment by HaloFour** [comment]
@theunrepentantgeek 

In this case the return value is a value type (since it's a tuple), but your point stands, it's impossible for `current` to be `null` so why would one want to treat it as a nullable-value-type?

**Comment by svick** [comment]
I think this would only be useful in cases where:

1. The type is not a reference type. For those, `var` is already nullable, with the null state being tracked, so that in the common case where `null` doesn't get involved later on, it behaves as if it the variable was not nullable.
2. You initially assign a non-nullable value, but want to assign `null` or a nullable value to the same variable afterwards.
3. The type name is sufficiently long.

As such, to me this seems like it wouldn't be useful that often. And in cases where it would be useful, you could use an extension method instead:

```c#
public static T? AsNullable<T>(this T value) where T : struct 
    => value;

var current = stack.Peek().AsNullable();
```
